### PR TITLE
Feature add --maxtextlength option, add xlunchr script, and fixes in genentries

### PR DESCRIPTION
**Pull request by sc0ttj**
added option `--maxtextlength`

added modified & fixed `xlunchr`

fixes in `genentries`, plus 2 new SVG to PNG methods (rsvgp-convert and inkscapelite)..
(could prob extend to use 'inkscape' as well as 'inkscapelite') 

TO DO (probably): 

* fix stuff below based on comments
* add xlunchr add assets for installation to /usr/share/xlunchr (button icons, themes, etc)
* update xlunchr to use the new dir above
* fix xlunchr to use to Noto* fonts it was bundled with

### CODE REVIEW

File: extra/genentries [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
-svgpath="svgicons"
+# this file generates an 'entries.dsv' for xlunch, which when 
+# generated, will contain lines like:
+#
+#  DeaDBeeF Music Player;/opt/deadbeef/deadbeef.png;/usr/bin/deadbeef
+
+# Usage:   genentries > /root/.config/xlunch/entries.dsv 
+
+# where to put the PNG icons we generate from svg
+svgpath="/usr/share/xlunch/svgicons"
```
Comment (PMunch) [inline extra/genentries:11]: Why did you change this? The script is intended to drop the svgs into a local directory if no argument is given and some installation scripts rely on this behaviour.

File: extra/genentries [hunk 2]
```diff
@@ -187,26 +227,62 @@ function readdesktopfile {
...
-          icon=$(find -L "/usr/share/icons" -type f \( "${findargs[@]}" -false \) -print -quit)
+          icon=$(find -L "/usr/share/icons" -type f \( "${findargs[@]}" -false \) -print -quit 2>/dev/null)
+        fi
+        if [ -z "$icon" ] && [ -d "/usr/share/pixmaps/puppy" ]; then
+          icon=$(find -L "/usr/share/pixmaps/puppy" -type f \( "${findargs[@]}" -false \) -print -quit 2>/dev/null)
         fi
       fi
 
+      # if the icon is a SVG, xlunch cant show it, so convert to PNG and
+      # put it in /usr/share/xlunch/svgicons/<name>.png
```
Comment (PMunch) [inline extra/genentries:247]: This path can be changed by a runtime switch, don't name a path here.

File: extra/genentries [hunk 3]
```diff
@@ -187,26 +227,62 @@ function readdesktopfile {
...
+            convert \
+              -density 500 \
+              -resize "$iconsize"x"$iconsize" \
+              -background none \
+              "$icon"
+          fi
         fi
-        icon="$svgpath/"$iconname".png"
+        # if a new PNG was created successfully, set it as our application icon
+        [ -f "${svgpath}/${iconname}.png" ] && icon="${svgpath}/${iconname}.png"
```
Comment (PMunch) [inline extra/genentries:285]: Not sure if this is the behaviour we want. If the file was not created we should still use  the invalid path to let users know that it wasn't able to convert the icon.

File: extra/xlunchr [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+# any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (PMunch) [inline extra/xlunchr:26]: This could maybe be rewritten? Right now it seems like it's a separate project, but this would merge it into xlunch.

File: xlunch.c [hunk 5]
```diff
@@ -1957,7 +1965,7 @@ void init(int argc, char **argv)
         exit(WINERROR);
     }
 
-    XMatchVisualInfo(disp, DefaultScreen(disp), 32, TrueColor, &vinfo);
+    XMatchVisualInfo(disp, DefaultScreen(disp), 24, TrueColor, &vinfo);
```
Comment (PMunch) [inline xlunch.c:1968]: Please change this back to support translucent backgrounds and background colours.

File: xlunch.c [hunk 6]
```diff
@@ -2253,18 +2261,38 @@ void renderEntry(Imlib_Image buffer, char title[256], node_t * current, Cursor *
...
-        } while(text_w > cell_width-(text_after ? (icon_size != 0 ? icon_padding*2 : icon_padding) + icon_size + text_padding : 2*text_padding) && sz>0);
+
+        if (maxtextlength >= 0) {
+          // get full title (--maxtextlength 0)
+          title = current->title;
+          extra_chars = 0;
+          // if value greater than zero, truncate text
+          if (maxtextlength > 0){
+            if (strlen(title) > maxtextlength) {
+              extra_chars = strlen(title)-maxtextlength;
```
Comment (PMunch) [inline xlunch.c:2277]: Since you're adding two dots after the title when truncated shouldn't you take those into account here? This risks making the text string longer if `strlen(title)` is only a single character longer than `maxtextlength`.

**Review by PMunch — CHANGES_REQUESTED** [review]
Overall looks pretty good. A couple of things I'm not to sure about. Please see comments.

File: extra/genentries (hunk 1 above)
Comment (sc0ttj) [reply extra/genentries:11]: Not a biggie ... but why have them dotted around in local dirs when you can keep all the xlunch assets in /usr/share/xlunch (which is the best place for them)? .. 

And personally , I didn't want an `svgicons` dir in my $HOME dir, and any OS with many users would get many duplicate folders..

File: extra/genentries (hunk 2 above)
Comment (sc0ttj) [reply extra/genentries:247]: :+1:

File: extra/xlunchr (hunk 4 above)
Comment (sc0ttj) [reply extra/xlunchr:26]: Maybe some additional info to say it was adapted from the original `xlunchr`?

File: extra/genentries (hunk 3 above)
Comment (sc0ttj) [reply extra/genentries:285]: Your call, I have it there cos that is what works for me to not get any missing icons..

File: xlunch.c (hunk 5 above)
Comment (sc0ttj) [reply xlunch.c:1968]: We might need a switch or check here then, as it won't build on my system (or any other recent Puppy Linux) unless it's 24 ...   

Not sure what the check would be.. 

See here for more info: http://murga-linux.com/puppy/viewtopic.php?p=998515#998515

File: xlunch.c (hunk 6 above)
Comment (sc0ttj) [reply xlunch.c:2277]: Maybe.. I don't know C ..  But I give `--maxtextlength 10`, and I get 10 chars before the `..` ... and if I give `--maxtextlength 1`, I get 1 char before the `..` - which is what I would expect...

File: extra/xlunchr (hunk 4 above)
Comment (sc0ttj) [reply extra/xlunchr:26]: ..or rename it to **something** like `xlunch-profile`, then remove the preceding `--` from the options, then you get usage like:

```
xlunch-profile desktop  # launch xlunch as desktop

xlunch-profile menu     # launch xlunch as menu

# etc...
```

Comment (sc0ttj) [reply extra/xlunchr:26]: I was in the process of making xlunchr source a config file, which would contain stuff like

```shell
XLUNCHR_DESKTOP_ICON_SIZE=48
XLUNCHR_DESKTOP_COLUMNS=4

XLUNCHR_MENU_ICON_SIZE=48
XLUNCHR_MENU_COLUMNS=4

# etc
```
.. that way users could define some ENV vars in `.bashrc` or wherever, and customise the various ready-made profile very easily, without affecting the default `xlunch` settings...

File: extra/genentries (hunk 1 above)
Comment (PMunch) [reply extra/genentries:11]: Well the idea is that you can generate them when testing but without overriding your actual setup. You can also have multiple folders for different things. All of this could of course be achieved by passing the argument, but there are currently things  that depend on this which would break by this change.

File: extra/genentries (hunk 3 above)
Comment (PMunch) [reply extra/genentries:285]: Well, missing icons because of an error when converting shouldn't be an accepted outcome. It should create an error so the user is aware of the issue.

File: extra/xlunchr (hunk 4 above)
Comment (PMunch) [reply extra/xlunchr:26]: Why not just have two config files for xlunch and let the users modify those?

File: xlunch.c (hunk 5 above)
Comment (PMunch) [reply xlunch.c:1968]: Hmm, that's a shame. Is it right that it doesn't build? Or that it just crashes at runtime as that post seems to indicate? If it's the latter it seems like an easy check to see if `XMatchVisualInfo` returns non-nil and then try it again with 24.

File: xlunch.c (hunk 6 above)
Comment (PMunch) [reply xlunch.c:2277]: Ah, if that is what you expect then it's okay. I would think `maxtextlength` would mean that if my text is longer than `maxtextlength` it would cut it down to fit within `maxtextlength`. In this case setting `--maxtextlength 1` would mean every string would be converted to be three characters long.

May I ask what you're using this option for by the way?

File: xlunch.c (hunk 5 above)
Comment (sc0ttj) [reply xlunch.c:1968]: crashed at runtime...

File: xlunch.c (hunk 6 above)
Comment (sc0ttj) [reply xlunch.c:2277]: cos i wanted to disable auto truncation and wasn't having any luck with negative text padding, or any other options... 

ideally, I would like text width to fill the space by default .. on my system, if I use `xlunch --columns 4`, I have *lots* of spacing between columns, but still very short text titles, truncated at about 5th char (not inc `..`) .. which is not long enough to even work out the prog names..  Ideally I wouldn't have to supply any options - xlunch would just allow for longer titles by default..  or something!

File: extra/xlunchr (hunk 4 above)
Comment (sc0ttj) [reply extra/xlunchr:26]: Can leave the ENV var stuff ... You guys would need to think about if you want xlunchr in this repo ... currently, this PR is missing its icons and themes, and IIRC, I deleted all the custom config stuff (replaced with a call to genentries), but can't remember how you choose xlunchr themes (I didn't play with that)... In this PR, that eature might be missing/broken...

Comment (sc0ttj) [reply extra/xlunchr:26]: Selecting diff background and button themes/icons should be done in xlunch options or genentries options anyway I guess... Up to you guys,,

File: extra/genentries (hunk 3 above)
Comment (sc0ttj) [reply extra/genentries:285]: So, remove the -f check and always set the var name?

File: extra/xlunchr (hunk 4 above)
Comment (PMunch) [reply extra/xlunchr:26]: Oh sorry, I thought you was the original creator of xlunchr. In that case we don't want to merge it into xlunch. It's a stand-alone project, and it might get merged in the future. But we'd want to do it properly with full integration.

File: xlunch.c (hunk 5 above)
Comment (PMunch) [reply xlunch.c:1968]: Okay, in that case it should just be a case of checking the return value of `XMatchVisualInfo`, if it's not 0 then it should be called again with 24 as  the depth. Could you try this and see if it works?

File: xlunch.c (hunk 6 above)
Comment (PMunch) [reply xlunch.c:2277]: Ah okay, have you seen [this](https://github.com/Tomas-M/xlunch/blob/master/docs/spacings.svg) by the way? What you describe should be possible to achieve with large `iconpadding` and small `textpadding`.

File: extra/xlunchr (hunk 4 above)
Comment (sc0ttj) [reply extra/xlunchr:26]: 👍

File: xlunch.c (hunk 6 above)
Comment (sc0ttj) [reply xlunch.c:2277]: I had not ... WIll have a play with it..

File: xlunch.c (hunk 5 above)
Comment (sc0ttj) [reply xlunch.c:1968]: I will try it when I get home

File: extra/xlunchr (hunk 4 above)
Comment (sc0ttj) [reply extra/xlunchr:26]: OK, I will remove and rename xlunchr, create my own repo for it, cheers

File: xlunch.c (hunk 6 above)
Comment (fredx181) [reply xlunch.c:2277]: The problem with setting `--iconpadding` to high value (to get larger text) is that it will also affect vertical padding, setting e.g. `--leastvmargin 1` doesn't help.
The option from sc0ttj `--maxtextlength` combined with high value of `--leastmargin` (and low value for `--iconpadding`) makes things more flexible for me.

Comment (PMunch) [reply xlunch.c:2277]: Well of course, you must allow more space for the text. I just tried @sc0ttj's solution and it has a couple of issues. Since no more space is required the text from one entry can overflow onto another entry, and the highlight around an icon might not cover all the text. I'm not sure if that's acceptable.. Could one of you give me an example the kind of layout you wish to create with this?

EDIT: Just realised you were talking about _vertical_ padding. Yeah I can see that being an issue, maybe adding a `--iconvpadding` option like `--leastmargin` has `--leastvmargin` would  be a better way to fix this?

Comment (PMunch) [reply xlunch.c:2277]: I just created this: https://github.com/Tomas-M/xlunch/commit/1a0dfcc242ccde3c96ee6205e2abd8d365eefa2d it adds a new parameter `--iconvpadding` which changes the vertical padding of icons. So instead of `--maxtextlength` you can now use `--iconpadding <large number> --iconvpadding <10, or another low number>`. This would extend the highlight over the entire entry so the text won't be outside the highlight, and it prevents text from flowing over other text, while still allowing longer text entries with small vertical padding.

Comment (fredx181) [reply xlunch.c:2277]: > So instead of --maxtextlength you can now use --iconpadding <large number> --iconvpadding <10, or another low number>.  
 
The `--iconvpadding` is a great addition, I can confirm that it's working well, thanks !

File: xlunch.c (hunk 5 above)
Comment (PMunch) [reply xlunch.c:1968]: Just fixed this in https://github.com/Tomas-M/xlunch/commit/85d4163a5eebdd5ba49e85cd80496d0a8fc24ceb, could you test that and see if it works for you?
